### ISSUE TITLE
Has the verb-attention implemented yet?

### ISSUE THREAD

**Issue by ehsan-soe** [pytorch/fairseq#1141]
Hi,
Thanks for the nice source.
I wonder if verb-attention proposed in [here](https://www.aclweb.org/anthology/P19-1254) implemented yet? or if there is a plan for it to be implemeneted soon?

**Comment by huihuifan** [comment]
thanks for your comment. We are working on this. If you would like to use it immediately, you can implement it by modifying the mask created in multi-head self-attention. An important note is: in our work, each verb was already pre-identified using semantic role labeling, and then flagged with a special <V> token to indicate the subsequent word was a verb. Thus, it was very easy for us to apply the verb attention mask by only allowing the attention to look at words after <V>. There is only one nuance: at the very beginning of the sequence, before any verbs have been seen, the attention mask cannot be 0 everywhere, or the model can cheat! It is very important at the first timestep to make sure the model is not cheating. We did this by having the model attention to the 0 vector. 

If your verbs are not pre-tagged like this, the attention mechanism will not know which words are verbs. Your setup needs to provide this information somehow - you could modify the batching for example to include which words are verbs if you don't want to modify the dataset to include <V>. But either way you will need to run a part of speech tagger or have a keyword list.

**Comment by huihuifan** [comment]
ah the comment does not look right on github. In our work, we tagged each verb with a [v] symbol, so the text looked like this:
[v] went [arg0] etc etc 
is what I meant.

**Comment by ehsan-soe** [comment]
@huihuifan Thanks for the nice explanation

**Comment by ehsan-soe** [comment]
Hi,
Any update on the code release for this nice paper?
Also it would be very nice if you could release the SRL decomposed version of writingprompt dataset? It would be really helpful since it is computationally expensive.

**Comment by huihuifan** [comment]
I have gotten sidetracked on some other things. But let me describe what I did so it's easier to implement- 

In fairseq/modules/MultiheadAttention, the attention heads are all computed in parallel because it's computationally efficient. To make it easier to experiment with various configurations of attention heads, I added the verb-attention mechanism in the downsampled_multihead_attention.py file, which computes the heads separately by calling the SingleHeadAttention mechanism. For verb-attention, we want n-1 of the attention heads to be "standard attention heads" and 1 head to be a "verb attention head" (though we experimented with other configurations. To make this modification, I arbitrarily assigned the 1st attention head to be the verb attention head. 

Then, when each attention head is called, for the verb attention head I pass in a flag that indicates the special verb token in the dictionary, for example [V]. Then before computing the attention, essentially you want to compute the verb mask by building a tensor that can only attend to words immediately after this special verb token (note that in our work, we modeled full vocabulary, not BPE... with BPE I suppose you would need to do some additional work here). And then put -inf everywhere else (unless it's the first timestep - if everything is -inf the model can cheat).  

for the SRL decomposition release- unfortunately, no :(  I'm sorry, but we are actually not allowed to release this modified version of the data. I know it's quite computationally intensive.

**Comment by ehsan-soe** [comment]
> I have gotten sidetracked on some other things. But let me describe what I did so it's easier to implement-
> 
> In fairseq/modules/MultiheadAttention, the attention heads are all computed in parallel because it's computationally efficient. To make it easier to experiment with various configurations of attention heads, I added the verb-attention mechanism in the downsampled_multihead_attention.py file, which computes the heads separately by calling the SingleHeadAttention mechanism. For verb-attention, we want n-1 of the attention heads to be "standard attention heads" and 1 head to be a "verb attention head" (though we experimented with other configurations. To make this modification, I arbitrarily assigned the 1st attention head to be the verb attention head.
> 
> Then, when each attention head is called, for the verb attention head I pass in a flag that indicates the special verb token in the dictionary, for example [V]. Then before computing the attention, essentially you want to compute the verb mask by building a tensor that can only attend to words immediately after this special verb token (note that in our work, we modeled full vocabulary, not BPE... with BPE I suppose you would need to do some additional work here). And then put -inf everywhere else (unless it's the first timestep - if everything is -inf the model can cheat).
> 
> for the SRL decomposition release- unfortunately, no :( I'm sorry, but we are actually not allowed to release this modified version of the data. I know it's quite computationally intensive.

Thanks @huihuifan 
Can you please correct me if I am wrong?
so I should create a verb mask in forward path of ```FConvDecoder``` using ```prev_output_tokens``` and the special verb token index. Then pass this verb mask to forward path of ```SelfAttention``` (kind of similar to ```key_padding_mask```, right?)

And this verb mask would be passed to ```DownsampledMultiHeadVerbAttention``` and will be used for the first head for example?

BTW, since during the training since we are using teacher forcing, the verb attention head is actually attending to ground truth previous verb, right?
